File: app/src/main/java/com/example/dronevision/utils/NGeoCalc.java
package com.example.dronevision.utils;

public class NGeoCalc {
    final double pza()  { return 6378136.0; }
    final double pzA()  { return 0.003352803735; } // 1.0 / 298.25784
    final double wgsa() { return 6378137.0; }
    final double wgsA() { return 0.003352810665; } // 1.0 / 298.257223563
    final double ska()  { return 6378245.0; }
    final double skA()  { return 0.003352329869; } // 1.0 / 298.3
    //
    public void geoToGeo(double[] Bb, double[] Lb, double[] Hb, double ab, double Ab,
                         double  Ba, double  La, double  Ha, double aa, double Aa,
                         double dx, double dy, double dz, double wx, double wy, double wz, double m) {
        double[] dB = new double[1];
        double[] dL = new double[1];
        double[] dH = new double[1];
        double B, L, H;
        geoToGeoDelta(dB, dL, dH, ab, Ab,
                Ba, La, Ha, aa, Aa,
                dx, dy, dz, wx, wy, wz, m);
        B = Ba + degreesToRadians(dB[0] / 3600) / 2;
        L = La + degreesToRadians(dL[0] / 3600) / 2;
        H = Ha + dH[0] / 2;
        geoToGeoDelta(dB, dL, dH, ab, Ab,
                B, L, H, aa, Aa,
                dx, dy, dz, wx, wy, wz, m);
        Bb[0] = Ba + degreesToRadians(dB[0] / 3600);
        Lb[0] = La + degreesToRadians(dL[0] / 3600);
        Hb[0] = Ha + dH[0];
    }
    //
    public void geoToGeoDelta(double[] dB, double[] dL, double[] dH, double ab, double Ab,
                              double  Ba, double  La, double  Ha, double aa, double Aa,
                              double dx, double dy, double dz, double wx, double wy, double wz, double m) {
        double da = ab - aa;
        double e2b = 2.0 * Ab - Ab * Ab;
        double e2a = 2.0 * Aa - Aa * Aa;
        double de2 = e2b - e2a;
        double a = (ab + aa) / 2.0;
        double e2 = (e2b + e2a) / 2.0;
        double sinB = Math.sin(Ba);
        double cosB = Math.cos(Ba);
        double cos2B = Math.cos(2 * Ba);
        double sinL = Math.sin(La);
        double cosL = Math.cos(La);
        double tanB = Math.tan(Ba);
        double M = a * (1 - e2) * Math.pow(1 - e2 * sinB * sinB, - 1.5);
        double N = a * Math.pow(1 - e2 * sinB * sinB, -0.5);
        double ro = 206264.806;
        dB[0] = ro / (M + Ha) *
                (
                        N / a * e2 * sinB * cosB * da
                                + (Math.pow(N, 2) / Math.pow(a, 2) + 1) * N * sinB * cosB * de2 / 2.0
                                - (dx * cosL + dy * sinL) * sinB + dz * cosB
                )
                - wx * sinL * (1 + e2 * cos2B)
                + wy * cosL * (1 + e2 * cos2B)
                - ro * m * e2 * sinB * cosB;
        dL[0] = ro / ((N + Ha) * cosB) * (-dx * sinL + dy * cosL)
                + tanB * (1 - e2) * (wx * cosL + wy * sinL)
                - wz;
        dH[0] = - a / N * da + N * sinB * sinB * de2 / 2
                + (dx * cosL + dy * sinL) * cosB
                + dz * sinB
                - N * e2 * sinB * cosB * (wx / ro * sinL - wy / ro * cosL)
                + (a * a / N + Ha) * m;

    }
    //
    double[] modf(double fullDouble) {
        int intVal = (int)fullDouble;
        double remainder = fullDouble - intVal;

        double[] retVal = new double[2];
        retVal[0] = intVal;
        retVal[1] = remainder;

        return retVal;
    }


    //
    public void geoToPlane(double[] x, double[] y, double B, double L, int nzone) {
        L = radiansToDegrees(positiveLongitude(L));
        double n;
        if (nzone > 0 && nzone <= 60) {
            n = nzone;
        } else {
            double[] ptr = modf(1.0 + L / 6.0);
            n = ptr[0];
        }
        double l = (L - (3.0 + 6.0 * (n - 1.0))) / 57.29577951;
        double l2 = l * l;
        double sinB = Math.sin(B);
        double sin2B = Math.sin(2 * B);
        double sinB2 = sinB * sinB;
        double sinB4 = sinB2 * sinB2;
        double sinB6 = sinB2 * sinB4;
        double cosB = Math.cos(B);
        x[0] = 6367558.4968 * B - sin2B * (16002.89 + 66.9607 * sinB2 + 0.3515 * sinB4 -
                l2 * (1594561.25 + 5336.535 * sinB2 + 26.79 * sinB4 + 0.149 * sinB6 +
                        l2 * (672483.4 - 811219.9 * sinB2 + 5420 * sinB4 - 10.6 * sinB6 +
                                l2 * (278194 - 830174 * sinB2 + 572434 * sinB4 - 16010 * sinB6 +
                                        l2 * (109500 - 574700 * sinB2 + 863700 * sinB4 - 398600 * sinB6)))));
        y[0] = (5 + 10 * n) * 100000 +
                l * cosB * (6378245 + 21346.1415 * sinB2 + 107.159 * sinB4 + 0.5977 * sinB6 +
                        l2 * (1070204.16 - 2136826.66 * sinB2 + 17.98 * sinB4 - 11.99 * sinB6 +
                                l2 * (270806 - 1523417 * sinB2 + 1327645 * sinB4 - 21701 * sinB6 +
                                        l2 * (79690 - 866190 * sinB2 + 1730360 * sinB4 - 945460 * sinB6))));
    }
    //
    public void planeToGeo(double[] B, double[] L, double x, double y) {
        double Beta = x / 6367558.4968;
        double sin2Beta = Math.sin(2 * Beta);
        double sinBeta = Math.sin(Beta);
        double sinBeta2 = sinBeta * sinBeta;
        double sinBeta4 = sinBeta2 * sinBeta2;
        double B0 = Beta + sin2Beta * (0.00252588685 - 0.0000149186 * sinBeta2 + 0.00000011904 * sinBeta4);
        double cosB0 = Math.cos(B0);
        double sin2B0 = Math.sin(2 * B0);
        double sinB0 = Math.sin(B0);
        double sinB02 = sinB0 * sinB0;
        double sinB04 = sinB02 * sinB02;
        double sinB06 = sinB04 * sinB02;
        double[] ptr = modf(y / 1000000.0);
        double n = ptr[0];
        double z0 = (y - (10.0 * n + 5.0) * 100000.0) / (6378245.0 * cosB0);
        double z02 = z0 * z0;
        double dB = - z02 * sin2B0 * (0.251684631 - 0.003369263 * sinB02 + 0.000011276 * sinB04
                - z02 * (0.10500614 - 0.04559916 * sinB02 + 0.00228901 * sinB04 - 0.00002987 * sinB06
                - z02 * (0.042858 - 0.025318 * sinB02 * 0.014346 * sinB04 - 0.001264 * sinB06
                - z02 * (0.01672 - 0.0063 * sinB02 + 0.01188 * sinB04 - 0.00328 * sinB06))));
        double l = z0 * (1 - 0.0033467108 * sinB02 - 0.0000056002 * sinB04 - 0.0000000187 * sinB06 -
                z02 * (0.16778975 + 0.16273586 * sinB02 - 0.0005249 * sinB04 - 0.00000846 * sinB06 -
                        z02 * (0.0420025 + 0.1487407 * sinB02 + 0.005942 * sinB04 - 0.000015 * sinB06 -
                                z02 * (0.01225 + 0.09477 * sinB02 + 0.03282 * sinB04 - 0.00034 * sinB06 -
                                        z02 * (0.0038 + 0.0524 * sinB02 + 0.0482 * sinB04 + 0.0032 * sinB06)))));
        B[0] = B0 + dB;
        L[0] = 6.0 * (n - 0.5) / 57.29577951 + l;
    }
    //
    final double dxSk42ToPz9002()              { return 23.93; }
    final double dySk42ToPz9002()              { return -141.03; }
    final double dzSk42ToPz9002()              { return -79.98; }
    final double wxGeoSk42ToPz9002()           { return 0.0; }
    final double wyGeoSk42ToPz9002()           { return -0.35; }
    final double wzGeoSk42ToPz9002()           { return -0.79; }
    final double wxSpatialSk42ToPz9002()       { return 0.0; }
    final double wySpatialSk42ToPz9002()       { return -0.0000016968; }
    final double wzSpatialSk42ToPz9002()       { return -0.00000383; }
    final double mSk42ToPz9002()               { return -0.00000022; }
    //
    final double dxPz9002ToWgs84()              { return -0.36; }
    final double dyPz9002ToWgs84()              { return 0.08; }
    final double dzPz9002ToWgs84()              { return 0.18; }
    final double wxGeoPz9002ToWgs84()           { return 0.0; }
    final double wyGeoPz9002ToWgs84()           { return 0.0; }
    final double wzGeoPz9002ToWgs84()           { return 0.0; }
    final double wxSpatialPz9002ToWgs84()       { return 0.0; }
    final double wySpatialPz9002ToWgs84()       { return 0.0; }
    final double wzSpatialPz9002ToWgs84()       { return 0.0; }
    final double mPz9002ToWgs84()               { return 0.0; }
    //
    double fmod(double x, double y) {
        double quot = x/y;
        return x - (quot < 0.0 ? Math.ceil(quot) : Math.floor(quot)) * y;
    }
    //
    double positiveLongitude(double value) {
        double pi2 = Math.PI * 2.0;
        double result = fmod(value, pi2);
        if (result < 0.0) {
            result = pi2 + result;
        }
        return result;
    }
    public static double degreesToRadians(double degrees) { return (Math.PI * degrees / 180.0); }
    public static double radiansToDegrees(double radians) { return (radians * 180.0 / Math.PI); }
    //
    public void wgs84ToPlane(double[] x, double[] y, double[] h, double B, double L, double H) {
        wgs84ToPlane(x, y, h, B, L, H, 0);
    }
    //
    public void wgs84ToPlane(double[] x, double[] y, double[] h, double B, double L, double H, int nzone) {
        NGeoCalc geoCalc = new NGeoCalc();
        double[] Bpz9002 = new double[1];
        double[] Lpz9002 = new double[1];
        double[] Hpz9002 = new double[1];
        double[] Bsk42   = new double[1];
        double[] Lsk42   = new double[1];
        double[] Hsk42   = new double[1];
        geoCalc.geoToGeo(Bpz9002, Lpz9002, Hpz9002, geoCalc.pza(), geoCalc.pzA(),
                B, L, H, geoCalc.wgsa(), geoCalc.wgsA(),
                -geoCalc.dxPz9002ToWgs84(), -geoCalc.dyPz9002ToWgs84(), -geoCalc.dzPz9002ToWgs84(),
                -geoCalc.wxGeoPz9002ToWgs84(), -geoCalc.wyGeoPz9002ToWgs84(), -geoCalc.wzGeoPz9002ToWgs84(),
                -geoCalc.mPz9002ToWgs84());
        geoCalc.geoToGeo(Bsk42, Lsk42, Hsk42, geoCalc.ska(), geoCalc.skA(),
                Bpz9002[0], Lpz9002[0], Hpz9002[0], geoCalc.pza(), geoCalc.pzA(),
                -geoCalc.dxSk42ToPz9002(), -geoCalc.dySk42ToPz9002(), -geoCalc.dzSk42ToPz9002(),
                -geoCalc.wxGeoSk42ToPz9002(), -geoCalc.wyGeoSk42ToPz9002(), -geoCalc.wzGeoSk42ToPz9002(),
                -geoCalc.mSk42ToPz9002());
        geoCalc.geoToPlane(x, y, Bsk42[0], Lsk42[0], nzone);
        h[0] = Hsk42[0];
    }
    //
    public void planeToWgs84(double[] B, double[] L, double[] H, double x, double y, double h) {
        NGeoCalc geoCalc = new NGeoCalc();
        double[] Bsk42   = new double[1];
        double[] Lsk42   = new double[1];
        double[] Bpz9002 = new double[1];
        double[] Lpz9002 = new double[1];
        double[] Hpz9002 = new double[1];
        //
        geoCalc.planeToGeo(Bsk42, Lsk42, x, y);
        geoCalc.geoToGeo(Bpz9002, Lpz9002, Hpz9002, geoCalc.pza(), geoCalc.pzA(),
                Bsk42[0], Lsk42[0], h, geoCalc.ska(), geoCalc.skA(),
                geoCalc.dxSk42ToPz9002(), geoCalc.dySk42ToPz9002(), geoCalc.dzSk42ToPz9002(),
                geoCalc.wxGeoSk42ToPz9002(), geoCalc.wyGeoSk42ToPz9002(), geoCalc.wzGeoSk42ToPz9002(),
                geoCalc.mSk42ToPz9002());
        geoCalc.geoToGeo(B, L, H, geoCalc.wgsa(), geoCalc.wgsA(),
                Bpz9002[0], Lpz9002[0], Hpz9002[0], geoCalc.pza(), geoCalc.pzA(),
                geoCalc.dxPz9002ToWgs84(), geoCalc.dyPz9002ToWgs84(), geoCalc.dzPz9002ToWgs84(),
                geoCalc.wxGeoPz9002ToWgs84(), geoCalc.wyGeoPz9002ToWgs84(), geoCalc.wzGeoPz9002ToWgs84(),
                geoCalc.mPz9002ToWgs84());
    }
    //
    public void planeToPlaneByZone(double[] xbuf, double[] ybuf, double[] hbuf,
                                   double x, double y, double h, int nzone) {
        double[] b = { 0 }, l = { 0 }, h2 = { 0 };
        planeToWgs84(b, l, h2, x, y, h);
        wgs84ToPlane(xbuf, ybuf, hbuf, b[0], l[0], h2[0], nzone);
    }
    //

}

File: app/src/main/java/com/example/dronevision/utils/NGeoCalc.kt
package com.example.dronevision.utils

class NGeoCalc {
    /* access modifiers changed from: package-private */
    fun pza(): Double {
        return 6378136.0
    }
    
    /* access modifiers changed from: package-private */
    fun pzA(): Double {
        return 0.003352803735
    }
    
    /* access modifiers changed from: package-private */
    fun wgsa(): Double {
        return 6378137.0
    }
    
    /* access modifiers changed from: package-private */
    fun wgsA(): Double {
        return 0.003352810665
    }
    
    /* access modifiers changed from: package-private */
    fun ska(): Double {
        return 6378245.0
    }
    
    /* access modifiers changed from: package-private */
    fun skA(): Double {
        return 0.003352329869
    }
    
    fun geoToGeo(
        Bb: DoubleArray,
        Lb: DoubleArray,
        Hb: DoubleArray,
        ab: Double,
        Ab: Double,
        Ba: Double,
        La: Double,
        Ha: Double,
        aa: Double,
        Aa: Double,
        dx: Double,
        dy: Double,
        dz: Double,
        wx: Double,
        wy: Double,
        wz: Double,
        m: Double
    ) {
        val dB = DoubleArray(1)
        val dL = DoubleArray(1)
        val dH = DoubleArray(1)
        geoToGeoDelta(dB, dL, dH, ab, Ab, Ba, La, Ha, aa, Aa, dx, dy, dz, wx, wy, wz, m)
        geoToGeoDelta(
            dB, dL, dH, ab, Ab, Ba + degreesToRadians(dB[0] / 3600.0) / 2.0, La + degreesToRadians(
                dL[0] / 3600.0
            ) / 2.0,
            Ha + dH[0] / 2.0, aa, Aa, dx, dy, dz, wx, wy, wz, m
        )
        Bb[0] = Ba + degreesToRadians(dB[0] / 3600.0)
        Lb[0] = La + degreesToRadians(dL[0] / 3600.0)
        Hb[0] = Ha + dH[0]
    }
    
    fun geoToGeoDelta(
        dB: DoubleArray,
        dL: DoubleArray,
        dH: DoubleArray,
        ab: Double,
        Ab: Double,
        Ba: Double,
        La: Double,
        Ha: Double,
        aa: Double,
        Aa: Double,
        dx: Double,
        dy: Double,
        dz: Double,
        wx: Double,
        wy: Double,
        wz: Double,
        m: Double
    ) {
        val da = ab - aa
        val e2b = Ab * 2.0 - Ab * Ab
        val e2a = Aa * 2.0 - Aa * Aa
        val de2 = e2b - e2a
        val a = (ab + aa) / 2.0
        val e2 = (e2b + e2a) / 2.0
        val sinB = Math.sin(Ba)
        val cosB = Math.cos(Ba)
        val cos2B = Math.cos(Ba * 2.0)
        val sinL = Math.sin(La)
        val cosL = Math.cos(La)
        val tanB = Math.tan(Ba)
        val d2 = e2b
        val M = (1.0 - e2) * a * Math.pow(1.0 - e2 * sinB * sinB, -1.5)
        val N = Math.pow(1.0 - e2 * sinB * sinB, -0.5) * a
        val d3 = e2a
        dB[0] =
            206264.806 / (M + Ha) * (N / a * e2 * sinB * cosB * da + (Math.pow(N, 2.0) / Math.pow(
                a,
                2.0
            ) + 1.0) * N * sinB * cosB * de2 / 2.0 - (dx * cosL + dy * sinL) * sinB + dz * cosB) - wx * sinL * (e2 * cos2B + 1.0) + wy * cosL * (e2 * cos2B + 1.0) - 206264.806 * m * e2 * sinB * cosB
        dL[0] =
            206264.806 / ((N + Ha) * cosB) * (-dx * sinL + dy * cosL) + (1.0 - e2) * tanB * (wx * cosL + wy * sinL) - wz
        dH[0] =
            -a / N * da + N * sinB * sinB * de2 / 2.0 + (dx * cosL + dy * sinL) * cosB + dz * sinB - N * e2 * sinB * cosB * (wx / 206264.806 * sinL - wy / 206264.806 * cosL) + (a * a / N + Ha) * m
    }
    
    /* access modifiers changed from: package-private */
    fun modf(fullDouble: Double): DoubleArray {
        val intVal = fullDouble.toInt()
        val d = intVal.toDouble()
        java.lang.Double.isNaN(d)
        return doubleArrayOf(intVal.toDouble(), fullDouble - d)
    }
    
    fun geoToPlane(x: DoubleArray, y: DoubleArray, B: Double, L: Double, nzone: Int) {
        val n: Double
        val L2 = radiansToDegrees(positiveLongitude(L))
        n = if (nzone <= 0 || nzone > 60) {
            modf(L2 / 6.0 + 1.0)[0]
        } else {
            nzone.toDouble()
        }
        val l = (L2 - ((n - 1.0) * 6.0 + 3.0)) / 57.29577951
        val l2 = l * l
        val sinB = Math.sin(B)
        val sin2B = Math.sin(2.0 * B)
        val sinB2 = sinB * sinB
        val sinB4 = sinB2 * sinB2
        val sinB6 = sinB2 * sinB4
        val cosB = Math.cos(B)
        x[0] =
            6367558.4968 * B - (66.9607 * sinB2 + 16002.89 + 0.3515 * sinB4 - (5336.535 * sinB2 + 1594561.25 + 26.79 * sinB4 + 0.149 * sinB6 + (672483.4 - 811219.9 * sinB2 + 5420.0 * sinB4 - 10.6 * sinB6 + (278194.0 - 830174.0 * sinB2 + 572434.0 * sinB4 - 16010.0 * sinB6 + (109500.0 - 574700.0 * sinB2 + 863700.0 * sinB4 - 398600.0 * sinB6) * l2) * l2) * l2) * l2) * sin2B
        y[0] =
            (10.0 * n + 5.0) * 100000.0 + l * cosB * (21346.1415 * sinB2 + 6378245.0 + 107.159 * sinB4 + 0.5977 * sinB6 + (1070204.16 - 2136826.66 * sinB2 + 17.98 * sinB4 - 11.99 * sinB6 + (270806.0 - 1523417.0 * sinB2 + 1327645.0 * sinB4 - 21701.0 * sinB6 + (79690.0 - 866190.0 * sinB2 + 1730360.0 * sinB4 - 945460.0 * sinB6) * l2) * l2) * l2)
    }
    
    fun planeToGeo(B: DoubleArray, L: DoubleArray, x: Double, y: Double) {
        val Beta = x / 6367558.4968
        val sin2Beta = Math.sin(Beta * 2.0)
        val sinBeta = Math.sin(Beta)
        val sinBeta2 = sinBeta * sinBeta
        val B0 =
            (0.00252588685 - 1.49186E-5 * sinBeta2 + 1.1904E-7 * sinBeta2 * sinBeta2) * sin2Beta + Beta
        val cosB0 = Math.cos(B0)
        val sin2B0 = Math.sin(2.0 * B0)
        val sinB0 = Math.sin(B0)
        val sinB02 = sinB0 * sinB0
        val sinB04 = sinB02 * sinB02
        val sinB06 = sinB04 * sinB02
        val d = Beta
        val d2 = sin2Beta
        val n = modf(y / 1000000.0)[0]
        val z0 = (y - (10.0 * n + 5.0) * 100000.0) / (6378245.0 * cosB0)
        val z02 = z0 * z0
        B[0] =
            B0 + -z02 * sin2B0 * (0.251684631 - 0.003369263 * sinB02 + 1.1276E-5 * sinB04 - (0.10500614 - 0.04559916 * sinB02 + 0.00228901 * sinB04 - 2.987E-5 * sinB06 - (0.042858 - 0.025318 * sinB02 * 0.014346 * sinB04 - 0.001264 * sinB06 - (0.01672 - 0.0063 * sinB02 + 0.01188 * sinB04 - 0.00328 * sinB06) * z02) * z02) * z02)
        L[0] =
            (n - 0.5) * 6.0 / 57.29577951 + (1.0 - 0.0033467108 * sinB02 - 5.6002E-6 * sinB04 - 1.87E-8 * sinB06 - (0.16273586 * sinB02 + 0.16778975 - 5.249E-4 * sinB04 - 8.46E-6 * sinB06 - (0.1487407 * sinB02 + 0.0420025 + 0.005942 * sinB04 - 1.5E-5 * sinB06 - (0.09477 * sinB02 + 0.01225 + 0.03282 * sinB04 - 3.4E-4 * sinB06 - (0.0524 * sinB02 + 0.0038 + 0.0482 * sinB04 + 0.0032 * sinB06) * z02) * z02) * z02) * z02) * z0
    }
    
    /* access modifiers changed from: package-private */
    fun dxSk42ToPz9002(): Double {
        return 23.93
    }
    
    /* access modifiers changed from: package-private */
    fun dySk42ToPz9002(): Double {
        return -141.03
    }
    
    /* access modifiers changed from: package-private */
    fun dzSk42ToPz9002(): Double {
        return -79.98
    }
    
    /* access modifiers changed from: package-private */
    fun wxGeoSk42ToPz9002(): Double {
        return 0.0
    }
    
    /* access modifiers changed from: package-private */
    fun wyGeoSk42ToPz9002(): Double {
        return -0.35
    }
    
    /* access modifiers changed from: package-private */
    fun wzGeoSk42ToPz9002(): Double {
        return -0.79
    }
    
    /* access modifiers changed from: package-private */
    fun mSk42ToPz9002(): Double {
        return -2.2E-7
    }
    
    /* access modifiers changed from: package-private */
    fun dxPz9002ToWgs84(): Double {
        return -0.36
    }
    
    /* access modifiers changed from: package-private */
    fun dyPz9002ToWgs84(): Double {
        return 0.08
    }
    
    /* access modifiers changed from: package-private */
    fun dzPz9002ToWgs84(): Double {
        return 0.18
    }
    
    /* access modifiers changed from: package-private */
    fun wxGeoPz9002ToWgs84(): Double {
        return 0.0
    }
    
    /* access modifiers changed from: package-private */
    fun wyGeoPz9002ToWgs84(): Double {
        return 0.0
    }
    
    /* access modifiers changed from: package-private */
    fun wzGeoPz9002ToWgs84(): Double {
        return 0.0
    }
    
    /* access modifiers changed from: package-private */
    fun mPz9002ToWgs84(): Double {
        return 0.0
    }
    
    /* access modifiers changed from: package-private */
    fun fmod(x: Double, y: Double): Double {
        val quot = x / y
        return x - (if (quot < 0.0) Math.ceil(quot) else Math.floor(quot)) * y
    }
    
    /* access modifiers changed from: package-private */
    fun positiveLongitude(value: Double): Double {
        val result = fmod(value, 6.283185307179586)
        return if (result < 0.0) {
            result + 6.283185307179586
        } else result
    }
    
    @JvmOverloads
    fun wgs84ToPlane(
        x: DoubleArray,
        y: DoubleArray,
        h: DoubleArray,
        B: Double,
        L: Double,
        H: Double,
        nzone: Int = 0
    ) {
        val geoCalc = NGeoCalc()
        val Bpz9002 = DoubleArray(1)
        val Lpz9002 = DoubleArray(1)
        val Hpz9002 = DoubleArray(1)
        val Bsk42 = DoubleArray(1)
        val Lsk42 = DoubleArray(1)
        val Hsk42 = DoubleArray(1)
        geoCalc.geoToGeo(
            Bpz9002,
            Lpz9002,
            Hpz9002,
            geoCalc.pza(),
            geoCalc.pzA(),
            B,
            L,
            H,
            geoCalc.wgsa(),
            geoCalc.wgsA(),
            -geoCalc.dxPz9002ToWgs84(),
            -geoCalc.dyPz9002ToWgs84(),
            -geoCalc.dzPz9002ToWgs84(),
            -geoCalc.wxGeoPz9002ToWgs84(),
            -geoCalc.wyGeoPz9002ToWgs84(),
            -geoCalc.wzGeoPz9002ToWgs84(),
            -geoCalc.mPz9002ToWgs84()
        )
        geoCalc.geoToGeo(
            Bsk42,
            Lsk42,
            Hsk42,
            geoCalc.ska(),
            geoCalc.skA(),
            Bpz9002[0],
            Lpz9002[0],
            Hpz9002[0],
            geoCalc.pza(),
            geoCalc.pzA(),
            -geoCalc.dxSk42ToPz9002(),
            -geoCalc.dySk42ToPz9002(),
            -geoCalc.dzSk42ToPz9002(),
            -geoCalc.wxGeoSk42ToPz9002(),
            -geoCalc.wyGeoSk42ToPz9002(),
            -geoCalc.wzGeoSk42ToPz9002(),
            -geoCalc.mSk42ToPz9002()
        )
        geoCalc.geoToPlane(x, y, Bsk42[0], Lsk42[0], nzone)
        h[0] = Hsk42[0]
    }
    
    fun planeToWgs84(
        B: DoubleArray,
        L: DoubleArray,
        H: DoubleArray,
        x: Double,
        y: Double,
        h: Double
    ) {
        val geoCalc = NGeoCalc()
        val Bsk42 = DoubleArray(1)
        val Lsk42 = DoubleArray(1)
        val Bpz9002 = DoubleArray(1)
        val Lpz9002 = DoubleArray(1)
        val Hpz9002 = DoubleArray(1)
        geoCalc.planeToGeo(Bsk42, Lsk42, x, y)
        val dArr = Bsk42
        val dArr2 = Lsk42
        geoCalc.geoToGeo(
            Bpz9002,
            Lpz9002,
            Hpz9002,
            geoCalc.pza(),
            geoCalc.pzA(),
            Bsk42[0],
            Lsk42[0],
            h,
            geoCalc.ska(),
            geoCalc.skA(),
            geoCalc.dxSk42ToPz9002(),
            geoCalc.dySk42ToPz9002(),
            geoCalc.dzSk42ToPz9002(),
            geoCalc.wxGeoSk42ToPz9002(),
            geoCalc.wyGeoSk42ToPz9002(),
            geoCalc.wzGeoSk42ToPz9002(),
            geoCalc.mSk42ToPz9002()
        )
        geoCalc.geoToGeo(
            B,
            L,
            H,
            geoCalc.wgsa(),
            geoCalc.wgsA(),
            Bpz9002[0],
            Lpz9002[0],
            Hpz9002[0],
            geoCalc.pza(),
            geoCalc.pzA(),
            geoCalc.dxPz9002ToWgs84(),
            geoCalc.dyPz9002ToWgs84(),
            geoCalc.dzPz9002ToWgs84(),
            geoCalc.wxGeoPz9002ToWgs84(),
            geoCalc.wyGeoPz9002ToWgs84(),
            geoCalc.wzGeoPz9002ToWgs84(),
            geoCalc.mPz9002ToWgs84()
        )
    }
    
    fun planeToPlaneByZone(
        xbuf: DoubleArray,
        ybuf: DoubleArray,
        hbuf: DoubleArray,
        x: Double,
        y: Double,
        h: Double,
        nzone: Int
    ) {
        val b = doubleArrayOf(0.0)
        val l = doubleArrayOf(0.0)
        val h2 = doubleArrayOf(0.0)
        planeToWgs84(b, l, h2, x, y, h)
        wgs84ToPlane(xbuf, ybuf, hbuf, b[0], l[0], h2[0], nzone)
    }
    
    companion object {
        fun degreesToRadians(degrees: Double): Double {
            return 3.141592653589793 * degrees / 180.0
        }
        
        fun radiansToDegrees(radians: Double): Double {
            return 180.0 * radians / 3.141592653589793
        }
        
        @JvmStatic
        fun main(argv: Array<String>) {
            val geoCalc = NGeoCalc()
            val x = DoubleArray(1)
            val y = DoubleArray(1)
            val h = DoubleArray(1)
            geoCalc.wgs84ToPlane(
                x,
                y,
                h,
                degreesToRadians(58.2342),
                degreesToRadians(38.9109),
                135.0,
                -1
            )
            val printStream = System.out
            printStream.println(
                x[0].toInt().toString() + " " + y[0].toInt() + " " + h[0].toInt()
            )
            val b = DoubleArray(1)
            val l = DoubleArray(1)
            geoCalc.planeToWgs84(b, l, h, x[0], y[0], h[0])
            val printStream2 = System.out
            printStream2.println(
                radiansToDegrees(b[0])
                    .toString() + " " + radiansToDegrees(
                    l[0]
                ) + " " + h[0]
            )
            val str = " "
            geoCalc.planeToPlaneByZone(x, y, h, 6457472.0, 7494882.0, 135.0, 8)
            val printStream3 = System.out
            printStream3.println("To zone 8: " + x[0].toInt() + str + y[0].toInt() + str + h[0].toInt())
            geoCalc.planeToPlaneByZone(x, y, h, 6457472.0, 7494882.0, 135.0, 6)
            val printStream4 = System.out
            printStream4.println("To zone 6: " + x[0].toInt() + str + y[0].toInt() + str + h[0].toInt())
        }
    }
}
